### ISSUE TITLE
SSR is not supported

### ISSUE THREAD

**Issue by alirezamirian** [adrdilauro/angular-ab-tests#8]
Default services depend on `window.navigator` and `document` to detect crawler and manipulate cookies.
These stuff don't work on the server with current implementation.

My suggestion is to at least expose injection tokens for these services, to enable overriding them for server platform.

**Comment by adrdilauro** [comment]
Thank you for pointing this out, I'll come out with a proper solution soon

**Comment by tom94zoe** [comment]
+1

**Comment by adrdilauro** [comment]
@alirezamirian @tom94zoe 

First of all, sorry for the delay, I have been particularly busy at work (as usual).

Then, unfortunately I have never worked with cookies and user agent on an Angular SSR project. I made a proposal of change in https://github.com/adrdilauro/angular-ab-tests/pull/9, but I am not sure if it would fit your needs? 

In order to make the module work on SSR, you need to pass a new argument to the `forRoot` module, and this argument has to respect the following interface: 

```typescript
export interface AbTestSsrAbstraction {
  getCookie(): string;
  setCookie(cookieString: string);
  getUserAgent(): string;
}
```

Basically, you have to provide three functions that would work in SSR, one for setting a cookie, one for reading a cookie, one for returning user agent. The implementation of each function is totally up to you.

Please let me know if this kind of interface would suit your needs. If you want to try it, check out the branch of the pull request and open the demo, and customise the configuration in this point: https://github.com/adrdilauro/angular-ab-tests/pull/9/files#diff-943d612766987a9e74235d09703085e5R12

**Comment by alirezamirian** [comment]
Thank you.
I think this resolves the issue.
However, I think there are a couple of other approaches worth considering. 
If you want to leave server implementation unimplemented, you could get almost the same result as this PR suggests, with only exporting some injection tokens (namely `AB_TESTS_COOKIE_HANDLER_TOKEN` and `AB_TESTS_CRAWLER_DETECTOR_TOKEN`). I mean Angular already supports overriding injectables in libraries, by providing them in your own module. I think the `forRoot` here is maybe superfluous.

Another possible approach is to ship a separate `AbTestsServerModule` which depends on the defacto standard injectables for server environment, like [HTTP_REQUEST][1] in [@nguniversal/express-engine] [2].

[1]: https://github.com/angular/universal/blob/83f659777a1f7914c357d5c0eb72705b04dc94ea/modules/express-engine/tokens/tokens.ts#L11
[2]: https://github.com/angular/universal

**Comment by alirezamirian** [comment]
In fact I think I can submit a PR for the latter approach if you want.

**Comment by adrdilauro** [comment]
> I think the forRoot here is maybe superfluous.

Do you mean that it should be possible to pass the configurations by letting the user inject them, rather than with a `forRoot` method?

Actually it seems a better idea; when I developed the module it was 18 months ago, I had much less experience of Angular than now, back then `forRoot` seemed to me the only way to configure a module. 

Anyway, now that it's already up and running and people are using it, I don't want to mess up with the configuration.

Regarding your proposal, yes please, that would be great, you can open a pull request with the addition of the separate module, it sounds like a cleaner solution. I have never used SSR, if you have experience with it and can contribute it would be great

**Comment by alirezamirian** [comment]
@adrdilauro I'm working on the PR. I want to remove injection tokens for `AB_TESTS_CRAWLER_DETECTOR_TOKEN`, `AB_TESTS_CRAWLER_DETECTOR_TOKEN`, and `AB_TESTS_RANDOM_EXTRACTOR_TOKEN`, because they are provided by a class and classes are perfectly fine DI tokens. In fact, `AbTestsService` is already provided without extra injection tokens.

Does this look fine to you?

**Comment by adrdilauro** [comment]
Hi @alirezamirian, I just published the new version that includes your contribution, it's `1.3.1`.

Thank you for this very useful addition to the module.
